services/blockstorage/adapter/leveldb_persistence.go

```diff
@@ -12,28 +30,138 @@ type levelDbBlockPersistence struct {
 	blockWritten chan bool
 	blockPairs   []*protocol.BlockPairContainer
 	config       Config
+	reporting    instrumentation.BasicLogger
+	db           *leveldb.DB
+}
+
+type config struct {
+	name string
+}
+
+func (c *config) NodeId() string {
+	return c.name
+}
+
+func NewLevelDbBlockPersistenceConfig(name string) Config {
+	return &config{name: name}
 }
 
 func NewLevelDbBlockPersistence(config Config) BlockPersistence {
+	db, err := leveldb.OpenFile("/tmp/db", nil)
+
+	if err != nil {
+		instrumentation.GetLogger(instrumentation.String("component", "persistence")).Error("Could not instantiate leveldb", instrumentation.Error(err))
+		panic("Could not instantiate leveldb")
+	}
+
 	return &levelDbBlockPersistence{
 		config:       config,
 		blockWritten: make(chan bool, 10),
+		db:           db,
+		reporting:    instrumentation.GetLogger(),
 	}
 }
 
+func (bp *levelDbBlockPersistence) WithLogger(reporting instrumentation.BasicLogger) BlockPersistence {
+	bp.reporting = reporting
+	return bp
+}
+
 func (bp *levelDbBlockPersistence) WriteBlock(blockPair *protocol.BlockPairContainer) {
-	bp.blockPairs = append(bp.blockPairs, blockPair)
+	var errors []error
+	var keys []string
+
+	if !basicValidation(blockPair) {
+		bp.reporting.Info("Block is invalid", instrumentation.Stringable("txBlockHeader", blockPair.TransactionsBlock.Header))
+		return
+	}
+
+	lastBlockHeight, lastBlockHeightRetrivalError := bp.loadLastBlockHeight()
```

Comment:
I couldn't understand from a brief documentation of the https://github.com/syndtr/goleveldb library what is its threading model.. is it thread safe? do you need to maintain sync? does it support transactions? what happens when several threads at the same time try to run `WriteBlock`?